test/conformance/certificate/utils.go

```diff
@@ -0,0 +1,175 @@
+		}
+		secretName := cert.Spec.SecretName
+		err = clients.KubeClient.Kube.CoreV1().Secrets(test.ServingNamespace).Delete(secretName, &metav1.DeleteOptions{})
+		if err != nil {
+			t.Errorf("Error cleaning up Secret %s for Certificate %s: %v", secretName, cert.Name, err)
+		}
+
+	}
+}
+
+// WaitForCertificateSecret polls the status of the Secret for the provided Certificate
+// until it exists or the timeout is exceeded. It then validates its contents
+func WaitForCertificateSecret(client *test.Clients, cert *v1alpha1.Certificate, desc string) error {
+	span := logging.GetEmitableSpan(context.Background(), fmt.Sprintf("WaitForCertificateSecret/%s/%s", cert.Spec.SecretName, desc))
+	defer span.End()
+
+	waitErr := wait.PollImmediate(test.PollInterval, test.PollTimeout, func() (bool, error) {
+		secret, err := client.KubeClient.Kube.CoreV1().Secrets(test.ServingNamespace).Get(cert.Spec.SecretName, metav1.GetOptions{})
+
+		if apierrs.IsNotFound(err) {
+			return false, nil
+		} else if err != nil {
+			return true, err
+		}
+
+		block, _ := pem.Decode(secret.Data[corev1.TLSCertKey])
+		if block == nil {
+			return true, fmt.Errorf("Failed to decode PEM data")
+		}
+
+		certData, err := x509.ParseCertificate(block.Bytes)
+		if err != nil {
+			return true, fmt.Errorf("Failed to parse certificate: %w", err)
+		}
+
+		got := certData.DNSNames
+		want := cert.Spec.DNSNames
+		if !reflect.DeepEqual(want, got) {
+			return true, fmt.Errorf("Incorrect DNSNames in secret. Want %v, got %v", want, got)
+		}
+
+		return true, nil
+	})
+
+	if waitErr != nil {
+		return fmt.Errorf("secret %q not found: %w", cert.Spec.SecretName, waitErr)
+	}
+	return nil
+}
+
+// WaitForCertificateState polls the status of the Certificate called name from client
+// every PollInterval until inState returns `true` indicating it is done, returns an
+// error or PollTimeout. desc will be used to name the metric that is emitted to
+// track how long it took for name to get into the state checked by inState.
+func WaitForCertificateState(client *test.NetworkingClients, name string, inState func(r *v1alpha1.Certificate) (bool, error), desc string) error {
+	span := logging.GetEmitableSpan(context.Background(), fmt.Sprintf("WaitForCertificateState/%s/%s", name, desc))
+	defer span.End()
+
+	var lastState *v1alpha1.Certificate
+	waitErr := wait.PollImmediate(test.PollInterval, test.PollTimeout, func() (bool, error) {
+		var err error
+		lastState, err = client.Certificates.Get(name, metav1.GetOptions{})
+		if err != nil {
+			return true, err
+		}
+		return inState(lastState)
+	})
+
+	if waitErr != nil {
+		return fmt.Errorf("certificate %q is not in desired state, got: %+v: %w", name, lastState, waitErr)
+	}
+	return nil
+}
+
+func getCertClass(t *testing.T, client *test.Clients) string {
+	configNetworkCM, err := client.KubeClient.Kube.CoreV1().ConfigMaps(system.Namespace()).Get("config-network", metav1.GetOptions{})
```

Comment:
Should this also return error in line with the other methods in this file?

Reply:
Yeah, I think that's reasonable.